File: tests/test_propulsion.cc
#include "Integration.h"
#include "PressureController.h"
#include "Propulsion.h"
#include "Simulator.h"

#include <Eigen/Dense>

#include <cassert>
#include <cmath>
#include <memory>
#include <stdexcept>
#include <vector>

namespace {
bool near(double a, double b, double tolerance = 1e-8) {
    return std::abs(a - b) <= tolerance;
}

double splitStep(double value, double step) {
    // y' = y - 10y: exact diagonal solve proves the IMEX utility is generic
    // and independent from propulsion or a particular nonlinear solver.
    return rsim::imexAdditiveStep(
        value, 0.0, step,
        [](double, double y) { return y; },
        [](double, double y) { return -10.0 * y; },
        [](double, double rhs, double coefficient, const auto&) {
            return rhs / (1.0 + 10.0 * coefficient);
        });
}
}  // namespace

int main() {
    const rsim::ImexTableau ars = rsim::ImexTableau::ars222();
    ars.validate();
    assert(ars.c.size() == 3U);
    assert(ars.A_implicit[1][1] > 0.0);
    assert(ars.A_implicit[2][2] > 0.0);

    double coarse = 1.0;
    for (int i = 0; i < 10; ++i) coarse = splitStep(coarse, 0.01);
    double fine = 1.0;
    for (int i = 0; i < 20; ++i) fine = splitStep(fine, 0.005);
    const double exact = std::exp(-0.9);
    assert(std::abs(fine - exact) < std::abs(coarse - exact));

    const std::vector<double> pc{1.0e4, 5.0e6};
    const std::vector<double> mr{0.5, 4.0};
    const std::vector<double> temperature(4, 3400.0);
    const std::vector<double> cstar(4, 1500.0);
    const std::vector<double> cf(4, 1.7);
    const std::vector<double> isp(4, 260.0);
    auto table = std::make_shared<rsim::CombustionTableData>(
        pc, mr, temperature, cstar, cf, isp, 5.0, "N2O", "ethanol");
    assert(near(table->cstar(2.0e6, 2.0), 1500.0));
    bool extrapolation_rejected = false;
    try {
        static_cast<void>(table->cstar(6.0e6, 2.0));
    } catch (const std::out_of_range&) {
        extrapolation_rejected = true;
    }
    assert(extrapolation_rejected);

    auto ambient = std::make_shared<rsim::Ambient>("vacuum", 0.0);
    auto oxidizer = std::make_shared<rsim::Tank>(
        "oxidizer", 0.2, 1.0, 0.0, 900.0, 80.0, 3.0e6);
    auto fuel = std::make_shared<rsim::Tank>(
        "fuel", 0.2, 1.0, 1.2, 800.0, 70.0, 3.0e6);
    auto engine = std::make_shared<rsim::Engine>(
        "engine", table, 1.0e-3, 5.0, Eigen::Vector3d::UnitX(),
        Eigen::Vector3d(0.0, 0.1, 0.0), ambient, 5.0e5, 1.0);
    auto network = std::make_shared<rsim::Propulsion>();
    network->connect(oxidizer, engine, std::make_shared<rsim::Valve>(1.0e-7),
                     rsim::FluidPhase::Liquid, rsim::InletRole::Oxidizer);
    network->connect(fuel, engine, std::make_shared<rsim::Orifice>(1.0e-7),
                     rsim::FluidPhase::Liquid, rsim::InletRole::Fuel);
    network->finalize();
    const double mass_before = network->totalStoredMass();
    rsim::Simulator simulator(0.01);
    simulator.addModel(*network);
    simulator.addModel(*engine);
    simulator.step();
    assert(network->totalStoredMass() < mass_before);
    assert(engine->thrust() > 0.0);
    assert(engine->force().x() > 0.0);
    assert(engine->torqueAbout(Eigen::Vector3d::Zero()).z() < 0.0);
    assert(network->lastSolveSummary().find("converged") != std::string::npos);

    // The last partial step consumes only the remaining limiting propellant.
    // On the following step, the missing inlet flames out the engine natively.
    fuel->setPropellantMass(1.0e-8);
    simulator.step();
    assert(near(fuel->propellantMass(), 0.0, 1e-14));
    simulator.step();
    assert(engine->pressure() < 10.0 * 6894.757293168);
    assert(near(engine->oxidizerMassFlow(), 0.0));
    assert(near(engine->fuelMassFlow(), 0.0));
    assert(near(engine->thrust(), 0.0));
    assert(engine->force().isZero());

    rsim::RigidBody body(10.0, Eigen::Matrix3d::Identity());
    body.addVariableMassSource(*oxidizer);
    bool duplicate_rejected = false;
    try {
        body.addVariableMassSource(*oxidizer);
    } catch (const std::invalid_argument&) {
        duplicate_rejected = true;
    }
    assert(duplicate_rejected);
    body.update();
    assert(body.mass() > 10.0);

    rsim::Propulsion skeletons;
    auto junction = std::make_shared<rsim::Junction>("junction");
    skeletons.connect(oxidizer, junction, std::make_shared<rsim::Pump>());
    bool pump_rejected = false;
    try {
        skeletons.finalize();
    } catch (const std::logic_error&) {
        pump_rejected = true;
    }
    assert(pump_rejected);

    // One gas source may conservatively feed any number of tank branches.
    auto nitrogen_source = std::make_shared<rsim::Tank>(
        "shared-N2", 0.2, 1.0, 0.0, 1000.0, 0.0, 5.0e6, 293.15,
        0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.2, "none", "Nitrogen");
    auto target_a = std::make_shared<rsim::Tank>(
        "target-a", 0.2, 1.0, 0.0, 1000.0, 60.0, 1.0e6, 293.15,
        0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0, "water", "Nitrogen");
    auto target_b = std::make_shared<rsim::Tank>(
        "target-b", 0.2, 1.0, 0.0, 1000.0, 60.0, 1.5e6, 293.15,
        0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0, "water", "Nitrogen");
    auto target_closed = std::make_shared<rsim::Tank>(
        "target-closed", 0.2, 1.0, 0.0, 1000.0, 60.0, 2.0e6, 293.15,
        0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0, "water", "Nitrogen");
    auto branches = std::make_shared<rsim::Propulsion>("three-way-N2");
    branches->connect(nitrogen_source, target_a,
                      std::make_shared<rsim::Orifice>(2.0e-7),
                      rsim::FluidPhase::Gas);
    branches->connect(nitrogen_source, target_b,
                      std::make_shared<rsim::Orifice>(7.0e-8),
                      rsim::FluidPhase::Gas);
    branches->connect(nitrogen_source, target_closed,
                      std::make_shared<rsim::Valve>(1.0e-6, 0.0),
                      rsim::FluidPhase::Gas);
    branches->finalize();
    const double branch_mass_before = branches->totalStoredMass();
    const double branch_energy_before = nitrogen_source->pressurantInternalEnergy() +
        target_a->pressurantInternalEnergy() +
        target_b->pressurantInternalEnergy() +
        target_closed->pressurantInternalEnergy();
    const double source_before = nitrogen_source->pressurantMass();
    const double a_before = target_a->pressurantMass();
    const double b_before = target_b->pressurantMass();
    const double closed_before = target_closed->pressurantMass();
    rsim::Simulator branch_simulator(0.01);
    branch_simulator.addModel(*branches);
    branch_simulator.step();
    const double flow_a = branches->connectionMassFlow(0U);
    const double flow_b = branches->connectionMassFlow(1U);
    assert(flow_a > flow_b && flow_b > 0.0);
    assert(near(branches->connectionMassFlow(2U), 0.0));
    assert(near(target_a->pressurantMass() - a_before, 0.01 * flow_a, 1e-12));
    assert(near(target_b->pressurantMass() - b_before, 0.01 * flow_b, 1e-12));
    assert(near(nitrogen_source->pressurantMass() - source_before,
                -0.01 * (flow_a + flow_b), 1e-12));
    assert(near(target_closed->pressurantMass(), closed_before, 1e-12));
    assert(near(branches->totalStoredMass(), branch_mass_before, 1e-12));
    const double branch_energy_after = nitrogen_source->pressurantInternalEnergy() +
        target_a->pressurantInternalEnergy() +
        target_b->pressurantInternalEnergy() +
        target_closed->pressurantInternalEnergy();
    assert(near(branch_energy_after, branch_energy_before, 1e-7));

    // A declared connection also conserves a compatible reverse flow.
    auto reverse_low = std::make_shared<rsim::Tank>(
        "reverse-low", 0.2, 1.0, 0.0, 1000.0, 60.0, 1.0e6, 293.15,
        0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0, "water", "Nitrogen");
    auto reverse_high = std::make_shared<rsim::Tank>(
        "reverse-high", 0.2, 1.0, 0.0, 1000.0, 60.0, 4.0e6, 293.15,
        0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0, "water", "Nitrogen");
    rsim::Propulsion reverse("reverse-N2");
    reverse.connect(reverse_low, reverse_high,
                    std::make_shared<rsim::Orifice>(1.0e-7),
                    rsim::FluidPhase::Gas);
    reverse.finalize();
    const double reverse_total = reverse.totalStoredMass();
    const double low_before = reverse_low->pressurantMass();
    rsim::Simulator reverse_simulator(0.01);
    reverse_simulator.addModel(reverse);
    reverse_simulator.step();
    assert(reverse.connectionMassFlow(0U) < 0.0);
    assert(reverse_low->pressurantMass() > low_before);
    assert(near(reverse.totalStoredMass(), reverse_total, 1e-12));

    bool species_rejected = false;
    try {
        auto helium = std::make_shared<rsim::Tank>(
            "helium", 0.2, 1.0, 0.0, 1000.0, 60.0, 1.0e6, 293.15,
            0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0,
            "water", "Helium");
        rsim::Propulsion incompatible;
        incompatible.connect(nitrogen_source, helium,
                             std::make_shared<rsim::Orifice>(1.0e-7),
                             rsim::FluidPhase::Gas);
    } catch (const std::invalid_argument&) {
        species_rejected = true;
    }
    assert(species_rejected);

    const auto pressure_tank = [](const std::string& name, double pressure) {
        return std::make_shared<rsim::Tank>(
            name, 0.2, 1.0, 0.0, 1000.0, 60.0, pressure, 293.15,
            0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0,
            "water", "Nitrogen");
    };
    auto low_pressure_tank = pressure_tank("controller-low", 0.8e6);
    auto low_valve = std::make_shared<rsim::Valve>(1.0e-7, 0.0);
    rsim::BangBangPressureController open_controller(
        "open-low", low_pressure_tank, low_valve, 1.0e6, 5.0e4, false);
    open_controller.update();
    assert(open_controller.valveOpen());
    assert(near(low_valve->opening(), 1.0));

    auto high_pressure_tank = pressure_tank("controller-high", 1.2e6);
    auto high_valve = std::make_shared<rsim::Valve>(1.0e-7, 1.0);
    rsim::BangBangPressureController close_controller(
        "close-high", high_pressure_tank, high_valve, 1.0e6, 5.0e4, true);
    close_controller.update();
    assert(!close_controller.valveOpen());
    assert(near(high_valve->opening(), 0.0));

    auto deadband_tank = pressure_tank("controller-deadband", 1.0e6);
    auto memory_closed_valve = std::make_shared<rsim::Valve>(1.0e-7, 1.0);
    auto memory_open_valve = std::make_shared<rsim::Valve>(1.0e-7, 0.0);
    rsim::BangBangPressureController memory_closed(
        "memory-closed", deadband_tank, memory_closed_valve,
        1.0e6, 5.0e4, false);
    rsim::BangBangPressureController memory_open(
        "memory-open", deadband_tank, memory_open_valve,
        1.0e6, 5.0e4, true);
    memory_closed.update();
    memory_open.update();
    assert(!memory_closed.valveOpen());
    assert(memory_open.valveOpen());

    auto controlled_source = std::make_shared<rsim::Tank>(
        "controlled-shared-N2", 0.2, 1.0, 0.0, 1000.0, 0.0, 5.0e6,
        293.15, 0.0, 1.0, rsim::BlowdownMode::Polytropic, 1.0,
        "none", "Nitrogen");
    auto controlled_low = pressure_tank("controlled-low", 0.8e6);
    auto controlled_high = pressure_tank("controlled-high", 1.2e6);
    auto controlled_low_valve = std::make_shared<rsim::Valve>(1.0e-7, 0.0);
    auto controlled_high_valve = std::make_shared<rsim::Valve>(1.0e-7, 1.0);
    auto low_controller = std::make_shared<rsim::BangBangPressureController>(
        "low-branch-controller", controlled_low, controlled_low_valve,
        1.0e6, 5.0e4, false);
    auto high_controller = std::make_shared<rsim::BangBangPressureController>(
        "high-branch-controller", controlled_high, controlled_high_valve,
        1.0e6, 5.0e4, true);
    auto controlled_network = std::make_shared<rsim::Propulsion>(
        "controlled-branches");
    controlled_network->connect(controlled_source, controlled_low,
                                controlled_low_valve, rsim::FluidPhase::Gas);
    controlled_network->connect(controlled_source, controlled_high,
                                controlled_high_valve, rsim::FluidPhase::Gas);
    controlled_network->finalize();
    const double controlled_source_before = controlled_source->pressurantMass();
    rsim::Simulator controlled_simulator(0.01);
    controlled_simulator.addModel(*low_controller);
    controlled_simulator.addModel(*high_controller);
    controlled_simulator.addModel(*controlled_network);
    controlled_simulator.step();
    assert(low_controller->valveOpen());
    assert(!high_controller->valveOpen());
    assert(controlled_network->connectionMassFlow(0U) > 0.0);
    assert(near(controlled_network->connectionMassFlow(1U), 0.0));
    assert(near(controlled_source_before - controlled_source->pressurantMass(),
                0.01 * controlled_network->connectionMassFlow(0U), 1e-12));

    return 0;
}

File: python/tests/test_propulsion.py
"""Native propulsion integration, ownership, and table-copy tests."""

import gc
import tempfile
import unittest
import weakref

import numpy as np

import rsim


class PropulsionTests(unittest.TestCase):
    def make_table(self):
        pc = np.array([1.0e4, 5.0e6])
        mr = np.array([0.5, 4.0])
        shape = (pc.size, mr.size)
        temperature = np.full(shape, 3400.0, order="F")
        cstar = np.full(shape, 1500.0, order="F")
        cf = np.full(shape, 1.7, order="F")
        isp = np.full(shape, 260.0, order="F")
        reference = weakref.ref(cstar)
        table = rsim.CombustionTableData(
            pc, mr, temperature, cstar, cf, isp, 5.0, "N2O", "ethanol", False
        )
        del pc, mr, temperature, cstar, cf, isp
        gc.collect()
        self.assertIsNone(reference())
        self.assertAlmostEqual(table.cstar(1.0e6, 2.0), 1500.0)
        return table

    def test_network_vehicle_schedule_and_mass_coupling(self):
        table = self.make_table()
        ambient = rsim.Ambient(pressure=0.0)
        ox = rsim.Tank("ox", 0.2, 1.0, 0.0, 900.0, 80.0, 3.0e6)
        fuel = rsim.Tank("fuel", 0.2, 1.0, 1.2, 800.0, 70.0, 3.0e6)
        engine = rsim.Engine("engine", table, 1.0e-3, 5.0,
                             ambient=ambient,
                             application_point=np.array([0.0, 0.1, 0.0]),
                             initial_pressure=5.0e5)
        propulsion = rsim.Propulsion()
        propulsion.connect(ox, engine, rsim.Valve(1.0e-7),
                           upstream_phase=rsim.FluidPhase.LIQUID,
                           downstream_role=rsim.InletRole.OXIDIZER)
        propulsion.connect(fuel, engine, rsim.Orifice(1.0e-7),
                           upstream_phase=rsim.FluidPhase.LIQUID,
                           downstream_role=rsim.InletRole.FUEL)
        propulsion.finalize()

        world = rsim.Frame("world-prop", rsim.InertialStatus.INERTIAL)
        vehicle = rsim.Vehicle("rocket-prop", world, 10.0, np.eye(3),
                               models=[propulsion], force_torques=[engine])
        vehicle.rigid_body.add_variable_mass_source(ox)
        vehicle.rigid_body.add_variable_mass_source(fuel)
        stored = propulsion.total_stored_mass
        simulator = rsim.Simulator(0.01)
        simulator.add_vehicle(vehicle)
        simulator.step()

        self.assertLess(propulsion.total_stored_mass, stored)
        self.assertGreater(engine.thrust, 0.0)
        self.assertGreater(engine.force[0], 0.0)
        self.assertAlmostEqual(vehicle.rigid_body.mass,
                               10.0 + propulsion.total_stored_mass)
        self.assertIn("converged", propulsion.last_solve_summary)

    def test_topology_is_immutable_and_skeleton_fails(self):
        tank = rsim.Tank("tank", 0.2, 1.0, 0.0, 900.0, 80.0, 3.0e6)
        ambient = rsim.Ambient()
        network = rsim.Propulsion()
        network.vent(tank, ambient, rsim.Orifice(1.0e-8))
        network.finalize()
        with self.assertRaises(RuntimeError):
            network.vent(tank, ambient, rsim.Orifice(1.0e-8))

        invalid = rsim.Propulsion("invalid")
        invalid.connect(tank, rsim.Junction("j"), rsim.Pump())
        with self.assertRaises(RuntimeError):
            invalid.finalize()

    def test_mat_round_trip_and_flow_laws(self):
        pc = np.array([1.0e5, 2.0e6])
        mr = np.array([1.0, 3.0])
        values = np.arange(4.0).reshape((2, 2), order="F") + 1000.0
        setup = rsim.CombustionTable(
            pc, mr, values + 2000.0, values, np.full((2, 2), 1.6),
            np.full((2, 2), 250.0), 4.0, "oxidizer", "fuel", True,
        )
        with tempfile.TemporaryDirectory() as directory:
            path = f"{directory}/combustion.mat"
            setup.save_mat(path)
            loaded = rsim.CombustionTable.load_mat(path)
            self.assertEqual(loaded.oxidizer, "oxidizer")
            self.assertTrue(loaded.frozen_at_throat)
            self.assertAlmostEqual(loaded.native.cstar(pc[1], mr[0]),
                                   values[1, 0])

        high = rsim.Ambient("high", 5.0e5, 300.0)
        vacuum = rsim.Ambient("vacuum", 0.0, 300.0)
        medium = rsim.Ambient("medium", 1.0e5, 300.0)
        restriction = rsim.Orifice(2.0e-6)
        choked_vacuum = restriction.mass_flow(high, vacuum, rsim.FluidPhase.GAS)
        choked_medium = restriction.mass_flow(high, medium, rsim.FluidPhase.GAS)
        self.assertGreater(choked_vacuum, 0.0)
        self.assertAlmostEqual(choked_vacuum, choked_medium, places=12)
        self.assertLess(restriction.mass_flow(vacuum, high,
                                              rsim.FluidPhase.GAS), 0.0)

        liquid = rsim.Tank("liquid", 0.2, 1.0, 0.0, 1000.0, 100.0,
                           5.0e5)
        line = rsim.Line(1.0, 0.01, 1.0e-5, 1.0e-3)
        self.assertGreater(line.mass_flow(liquid, medium,
                                          rsim.FluidPhase.LIQUID), 0.0)

    def test_shared_pressurant_branches_conserve_mass_and_energy(self):
        def tank(name, pressure, pressurant="Nitrogen"):
            return rsim.Tank(
                name, 0.2, 1.0, 0.0, 1000.0, 60.0, pressure,
                temperature=293.15,
                blowdown_mode=rsim.BlowdownMode.POLYTROPIC,
                polytropic_exponent=1.0,
                propellant="water",
                pressurant=pressurant,
            )

        source = rsim.Tank(
            "shared-N2", 0.2, 1.0, 0.0, 1000.0, 0.0, 5.0e6,
            temperature=293.15,
            blowdown_mode=rsim.BlowdownMode.POLYTROPIC,
            polytropic_exponent=1.0,
            propellant="none",
            pressurant="Nitrogen",
        )
        targets = [tank("a", 1.0e6), tank("b", 1.5e6), tank("closed", 2.0e6)]
        network = rsim.Propulsion("branched-N2")
        network.connect(source, targets[0], rsim.Orifice(2.0e-7),
                        upstream_phase=rsim.FluidPhase.GAS)
        network.connect(source, targets[1], rsim.Orifice(7.0e-8),
                        upstream_phase=rsim.FluidPhase.GAS)
        network.connect(source, targets[2], rsim.Valve(1.0e-6, opening=0.0),
                        upstream_phase=rsim.FluidPhase.GAS)
        network.finalize()

        tanks = [source, *targets]
        masses_before = np.array([item.pressurant_mass for item in tanks])
        energy_before = sum(item.pressurant_internal_energy for item in tanks)
        total_before = network.total_stored_mass
        simulator = rsim.Simulator(0.01)
        simulator.add_model(network)
        simulator.step()

        flow_a = network.connection_mass_flow(0)
        flow_b = network.connection_mass_flow(1)
        self.assertGreater(flow_a, flow_b)
        self.assertGreater(flow_b, 0.0)
        self.assertEqual(network.connection_mass_flow(2), 0.0)
        self.assertAlmostEqual(targets[0].pressurant_mass - masses_before[1],
                               flow_a * 0.01, places=12)
        self.assertAlmostEqual(targets[1].pressurant_mass - masses_before[2],
                               flow_b * 0.01, places=12)
        self.assertAlmostEqual(source.pressurant_mass - masses_before[0],
                               -(flow_a + flow_b) * 0.01, places=12)
        self.assertEqual(targets[2].pressurant_mass, masses_before[3])
        self.assertAlmostEqual(network.total_stored_mass, total_before, places=12)
        self.assertAlmostEqual(
            sum(item.pressurant_internal_energy for item in tanks),
            energy_before,
            places=7,
        )

        reverse_low = tank("reverse-low", 1.0e6)
        reverse_high = tank("reverse-high", 4.0e6)
        reverse = rsim.Propulsion("reverse")
        reverse.connect(reverse_low, reverse_high, rsim.Orifice(1.0e-7),
                        upstream_phase=rsim.FluidPhase.GAS)
        reverse.finalize()
        reverse_total = reverse.total_stored_mass
        low_before = reverse_low.pressurant_mass
        reverse_simulator = rsim.Simulator(0.01)
        reverse_simulator.add_model(reverse)
        reverse_simulator.step()
        self.assertLess(reverse.connection_mass_flow(0), 0.0)
        self.assertGreater(reverse_low.pressurant_mass, low_before)
        self.assertAlmostEqual(reverse.total_stored_mass, reverse_total, places=12)

        with self.assertRaises(ValueError):
            incompatible = rsim.Propulsion("incompatible")
            incompatible.connect(source, tank("helium", 1.0e6, "Helium"),
                                 rsim.Orifice(1.0e-7),
                                 upstream_phase=rsim.FluidPhase.GAS)

        liquid_source = tank("liquid-source", 4.0e6)
        liquid_receiver = rsim.Tank(
            "liquid-receiver", 0.2, 1.0, 0.0, 1000.0, 10.0, 1.0e6,
            temperature=293.15,
            blowdown_mode=rsim.BlowdownMode.POLYTROPIC,
            polytropic_exponent=1.0,
            propellant="water",
            pressurant="Nitrogen",
        )
        liquid_network = rsim.Propulsion("liquid-transfer")
        liquid_network.connect(
            liquid_source, liquid_receiver, rsim.Orifice(1.0e-7),
            upstream_phase=rsim.FluidPhase.LIQUID,
        )
        liquid_network.finalize()
        liquid_total = liquid_network.total_stored_mass
        receiver_before = liquid_receiver.propellant_mass
        liquid_simulator = rsim.Simulator(0.01)
        liquid_simulator.add_model(liquid_network)
        liquid_simulator.step()
        self.assertGreater(liquid_receiver.propellant_mass, receiver_before)
        self.assertAlmostEqual(liquid_network.total_stored_mass,
                               liquid_total, places=12)

    def test_bang_bang_pressure_control_and_independent_branches(self):
        def tank(name, pressure):
            return rsim.Tank(
                name, 0.2, 1.0, 0.0, 1000.0, 60.0, pressure,
                temperature=293.15,
                blowdown_mode=rsim.BlowdownMode.POLYTROPIC,
                polytropic_exponent=1.0,
                propellant="water",
                pressurant="Nitrogen",
            )

        low, high, middle = (
            tank("low", 0.8e6), tank("high", 1.2e6), tank("middle", 1.0e6)
        )
        low_valve = rsim.Valve(1.0e-7, opening=0.0)
        high_valve = rsim.Valve(1.0e-7, opening=1.0)
        memory_open_valve = rsim.Valve(1.0e-7, opening=0.0)
        memory_closed_valve = rsim.Valve(1.0e-7, opening=1.0)
        controllers = [
            rsim.BangBangPressureController(
                "open-low", low, low_valve, 1.0e6, 5.0e4,
                initially_open=False,
            ),
            rsim.BangBangPressureController(
                "close-high", high, high_valve, 1.0e6, 5.0e4,
                initially_open=True,
            ),
            rsim.BangBangPressureController(
                "remember-open", middle, memory_open_valve, 1.0e6, 5.0e4,
                initially_open=True,
            ),
            rsim.BangBangPressureController(
                "remember-closed", middle, memory_closed_valve, 1.0e6, 5.0e4,
                initially_open=False,
            ),
        ]
        simulator = rsim.Simulator(0.01)
        for controller in controllers:
            simulator.add_model(controller)
        simulator.step()
        self.assertTrue(controllers[0].valve_open)
        self.assertFalse(controllers[1].valve_open)
        self.assertTrue(controllers[2].valve_open)
        self.assertFalse(controllers[3].valve_open)

        source = rsim.Tank(
            "shared-source", 0.2, 1.0, 0.0, 1000.0, 0.0, 5.0e6,
            temperature=293.15,
            blowdown_mode=rsim.BlowdownMode.POLYTROPIC,
            polytropic_exponent=1.0,
            propellant="none",
            pressurant="Nitrogen",
        )
        branch_network = rsim.Propulsion("controlled-branches")
        branch_network.connect(source, low, low_valve,
                               upstream_phase=rsim.FluidPhase.GAS)
        branch_network.connect(source, high, high_valve,
                               upstream_phase=rsim.FluidPhase.GAS)
        branch_network.finalize()
        source_before = source.pressurant_mass
        branch_simulator = rsim.Simulator(0.01)
        branch_simulator.add_model(controllers[0])
        branch_simulator.add_model(controllers[1])
        branch_simulator.add_model(branch_network)
        branch_simulator.step()
        self.assertGreater(branch_network.connection_mass_flow(0), 0.0)
        self.assertEqual(branch_network.connection_mass_flow(1), 0.0)
        self.assertAlmostEqual(
            source_before - source.pressurant_mass,
            branch_network.connection_mass_flow(0) * 0.01,
            places=12,
        )


if __name__ == "__main__":
    unittest.main()
